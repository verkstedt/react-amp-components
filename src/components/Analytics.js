/* eslint-disable react/forbid-prop-types, react/no-danger */
import React, { Component } from 'react'
import { object } from 'prop-types'
import { Helmet } from 'react-helmet'

class Analytics extends Component {
  static propTypes = {
    requests: object.isRequired,
    vars: object,
    extraUrlParams: object,
    triggers: object,
    transport: object
  }

  static defaultProps = {
    vars: {},
    extraUrlParams: {},
    triggers: {},
    transport: {}
  }

  render() {
    const {
      requests, vars, extraUrlParams, triggers, transport
    } = this.props
    const payload = {
      requests,
      vars,
      extraUrlParams,
      triggers,
      transport
    }
    return [
      <Helmet key="helmet">
        <script async="" custom-element="amp-analytics" src="https://cdn.ampproject.org/v0/amp-analytics-0.1.js" />
      </Helmet>,
      <amp-analytics>
        <script type="application/json" dangerouslySetInnerHTML={{ __html: JSON.stringify(payload) }} />
      </amp-analytics>
    ]
  }
}

export default Analytics
