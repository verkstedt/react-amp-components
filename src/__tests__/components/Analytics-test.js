/* eslint-disable no-template-curly-in-string */
import React from 'react'
import Analytics from '../../components/Analytics'
import Helmet from '../../utils/Helmet'
import { renderComponent } from '../test-utils'

const requests = {
  base: 'https://example.com/analytics?a=${account}&u=${canonicalUrl}&t=${title}',
  pageview: '${base}&type=pageview',
  event: '${base}&type=event&eventId=${eventId}'
}
const vars = {
  account: 'ABC123',
  countryCode: 'tr',
  tags: ['Swift,Jonathan', "Gulliver's Travels"]
}

const extraUrlParams = {
  a: '1',
  b: '2',
  c: '3'
}

const triggers = {
  sampledOnRandom: {
    on: 'visible',
    request: 'request',
    sampleSpec: {
      sampleOn: '${random}',
      threshold: 50
    }
  },
  sampledOnClientId: {
    on: 'visible',
    request: 'request',
    sampleSpec: {
      sampleOn: '${clientId(cookieName)}',
      threshold: 1
    }
  }
}

const transport = {
  beacon: false,
  xhrpost: false,
  image: true
}

describe('Analytics', () => {
  it('matches the snapshot', () => {
    const analytics = renderComponent(
      <Analytics {...{
        requests,
        vars,
        extraUrlParams,
        triggers,
        transport
      }}
      />
    )
    expect(analytics.toJSON()).toMatchSnapshot()
  })

  it('injects the right script tag', () => {
    renderComponent(
      <Analytics {...{
        requests,
        vars,
        extraUrlParams,
        triggers,
        transport
      }}
      />
    )
    Helmet.canUseDOM = false
    const staticHead = Helmet.renderStatic()
    expect(staticHead.scriptTags).toMatchSnapshot()
  })
})
