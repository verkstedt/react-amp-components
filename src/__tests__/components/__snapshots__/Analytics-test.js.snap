// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics injects the right script tag 1`] = `
Array [
  Object {
    "async": "",
    "custom-element": "amp-analytics",
    "src": "https://cdn.ampproject.org/v0/amp-analytics-0.1.js",
  },
]
`;

exports[`Analytics matches the snapshot 1`] = `
<amp-analytics>
  <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "{\\"requests\\":{\\"base\\":\\"https://example.com/analytics?a=\${account}&u=\${canonicalUrl}&t=\${title}\\",\\"pageview\\":\\"\${base}&type=pageview\\",\\"event\\":\\"\${base}&type=event&eventId=\${eventId}\\"},\\"vars\\":{\\"account\\":\\"ABC123\\",\\"countryCode\\":\\"tr\\",\\"tags\\":[\\"Swift,Jonathan\\",\\"Gulliver's Travels\\"]},\\"extraUrlParams\\":{\\"a\\":\\"1\\",\\"b\\":\\"2\\",\\"c\\":\\"3\\"},\\"triggers\\":{\\"sampledOnRandom\\":{\\"on\\":\\"visible\\",\\"request\\":\\"request\\",\\"sampleSpec\\":{\\"sampleOn\\":\\"\${random}\\",\\"threshold\\":50}},\\"sampledOnClientId\\":{\\"on\\":\\"visible\\",\\"request\\":\\"request\\",\\"sampleSpec\\":{\\"sampleOn\\":\\"\${clientId(cookieName)}\\",\\"threshold\\":1}}},\\"transport\\":{\\"beacon\\":false,\\"xhrpost\\":false,\\"image\\":true}}",
      }
    }
    type="application/json"
  />
</amp-analytics>
`;
